lib/log.c

```diff
@@ -202,42 +314,104 @@ static void vzlog_file(struct zlog *zl, struct timestamp_control *tsctl,
 	fflush(fp);
 }
 
+/* Search a buf for the filter strings */
+static int search_buf(const char *buf)
+{
+	char *found = NULL;
+
+	for (int i = 0; i < zlog_filter_count; i++) {
+		found = strstr(buf, zlog_filters[i]);
+		if (found != NULL)
+			return 0;
+	}
+
+	return -1;
+}
+
+/* Filter out a log */
+static int vzlog_filter(struct zlog *zl, struct timestamp_control *tsctl,
+			const char *proto_str, int priority, const char *format,
+			va_list args)
+{
+	int len = 0;
+	int ret = 0;
+	char buf[1024] = "";
+	va_list ac;
+
+	ret = time_print_buf(buf, len, sizeof(buf), tsctl);
+
+	len += ret;
+	if ((ret < 0) || ((size_t)len >= sizeof(buf)))
+		return search_buf(buf);
+
+	if (zl && zl->record_priority)
+		ret = snprintf(buf + len, sizeof(buf) - len,
+			       "%s: %s: ", zlog_priority[priority], proto_str);
+	else
+		ret = snprintf(buf + len, sizeof(buf) - len, "%s: ", proto_str);
+
+	len += ret;
+	if ((ret < 0) || ((size_t)len >= sizeof(buf)))
+		return search_buf(buf);
+
+	va_copy(ac, args);
+	ret = vsnprintf(buf + len, sizeof(buf) - len, format, ac);
+	va_end(ac);
+
+	len += ret;
+	if ((ret < 0) || ((size_t)len >= sizeof(buf)))
+		return search_buf(buf);
+
+	return search_buf(buf);
+}
+
 /* va_list version of zlog. */
 void vzlog(int priority, const char *format, va_list args)
 {
 	pthread_mutex_lock(&loglock);
 
-	char proto_str[32];
+	char proto_str[32] = "";
 	int original_errno = errno;
-	struct timestamp_control tsctl;
+	struct timestamp_control tsctl = {};
 	tsctl.already_rendered = 0;
 	struct zlog *zl = zlog_default;
 	char buf[256], *msg;
 
+	if (zl == NULL) {
+		tsctl.precision = 0;
+	} else {
+		tsctl.precision = zl->timestamp_precision;
+		if (zl->instance)
+			sprintf(proto_str, "%s[%d]: ", zl->protoname,
+				zl->instance);
+		else
+			sprintf(proto_str, "%s: ", zl->protoname);
+	}
+
+	/* If it doesn't match on a filter, do nothing with the log */
+	if (zlog_filter_count
+	    && vzlog_filter(zl, &tsctl, proto_str, priority, format, args)) {
```

Comment:
NAK on passing `format, args` here - that means we do the formatting twice, incurring twice the cost.  Just move up the `vasnprintfrr()` below if anything.